src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java

```diff
@@ -13,61 +14,148 @@
 import io.stargate.sgv2.jsonapi.service.operation.model.ModifyOperation;
 import io.stargate.sgv2.jsonapi.service.shredding.model.DocumentId;
 import io.stargate.sgv2.jsonapi.service.shredding.model.WritableShreddedDocument;
+import java.util.Collections;
 import java.util.List;
 import java.util.function.Supplier;
 
-/** Operation that inserts one or more documents. */
+/**
+ * Operation that inserts one or more documents.
+ *
+ * @param commandContext Context that defines namespace and database.
+ * @param documents Documents to insert.
+ * @param ordered If insert should be ordered.
+ */
 public record InsertOperation(
-    CommandContext commandContext, List<WritableShreddedDocument> documents)
+    CommandContext commandContext, List<WritableShreddedDocument> documents, boolean ordered)
     implements ModifyOperation {
 
   public InsertOperation(CommandContext commandContext, WritableShreddedDocument document) {
-    this(commandContext, List.of(document));
+    this(commandContext, List.of(document), true);
   }
 
   /** {@inheritDoc} */
   @Override
   public Uni<Supplier<CommandResult>> execute(QueryExecutor queryExecutor) {
+    if (ordered) {
+      return insertOrdered(queryExecutor);
+    } else {
+      return insertUnordered(queryExecutor);
+    }
+  }
+
+  // implementation for the ordered insert
+  private Uni<Supplier<CommandResult>> insertOrdered(QueryExecutor queryExecutor) {
+    // build query once
+    QueryOuterClass.Query query = buildInsertQuery();
+
+    return Multi.createFrom()
+        .iterable(documents)
+
+        // concatenate to respect ordered
+        .onItem()
+        .transformToUniAndConcatenate(
+            doc ->
+                insertDocument(queryExecutor, query, doc)
+
+                    // if we fail to insert propagate the FailFastInsertException
+                    .onFailure()
+                    .transform(t -> new FailFastInsertException(doc, t)))
+
+        // if no failures reduce to the op page
+        .collect()
+        .in(InsertOperationPage::new, (agg, in) -> agg.aggregate(in, null))
+
+        // use object identity to resolve to Supplier<CommandResult>
+        .map(i -> i)
+
+        // in case upstream propagated FailFastInsertException
+        // ensure to construct correctly the information about inserted documents
+        .onFailure(FailFastInsertException.class)
+        .recoverWithItem(
+            e -> {
+              // safe to cast, asserted class in onFailure
+              FailFastInsertException failFastInsertException = (FailFastInsertException) e;
+              WritableShreddedDocument failed = failFastInsertException.document;
+
+              // collect inserted, since it's sequential iterate until failed index
+              int failedIndex = documents().lastIndexOf(failed);
```

Comment:
Do we need this logic, since we use concatenate is already ordered and status can be captured with invoke() similar to unordered.

InsertOperationPage response = new InsertOperationPage();
`insertDocument(queryExecutor, query, doc)
                    .onItemOrFailure()
                    .invoke(
                        (documentId, error) -> {
                          response.aggregate(doc.id(), error);
                        }))`

Reply:
Hmm but how can I reduce? Where is the `response` coming from?

Reply:
@ivansenic 
`InsertOperationPage response = new InsertOperationPage();
    return Multi.createFrom()
        .items(documents.stream())
        .onItem()
        .transformToUniAndConcatenate(
            doc ->
                insertDocument(queryExecutor, query, doc)
                    .onItemOrFailure()
                    .invoke(
                        (documentId, error) -> {
                          response.aggregate(doc.id(), error);
                        }))
        .collect()
        .last()
        .onItemOrFailure()
        .transform(
            (id, error) -> {
              return response;
            });`

Reply:
Hmm, I don't like those side effects, especially when I am using a non-thread-safe structure.. Let me see if I can optimize this..

Reply:
@maheshrajamani this is the bug from the demo, I can not use last index on a document, as if first fails, I would assume that the first was written.. Need to change this :)